fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GroupByCountDistinctDataSkewEliminateRule.java

```diff
@@ -0,0 +1,172 @@
+        LogicalAggregationOperator aggOp = input.getOp().cast();
+        return aggOp.checkGroupByCountDistinctWithSkewHint() || (aggOp.checkGroupByCountDistinct() &&
+                context.getSessionVariable().isEnableDistinctColumnBucketization());
+    }
+
+    private static final GroupByCountDistinctDataSkewEliminateRule INSTANCE =
+            new GroupByCountDistinctDataSkewEliminateRule();
+
+    public static GroupByCountDistinctDataSkewEliminateRule getInstance() {
+        return INSTANCE;
+    }
+
+    // compute the type of bucket column, since bucket column introduce extra cost, so we
+    // choose just wide enough type to keep bucket number.
+    private Type pickBucketType(int bucketNum) {
+        Preconditions.checkArgument(0 < bucketNum && bucketNum <= 65536);
+        return (bucketNum <= 256) ? Type.TINYINT : Type.SMALLINT;
+    }
+
+    private ScalarOperator createBucketColumn(ColumnRefOperator distinctColumn, int bucketNum) {
+        CastOperator stringCol = new CastOperator(Type.VARCHAR, distinctColumn, true);
+        Function hashFunc = Expr.getBuiltinFunction(FunctionSet.MURMUR_HASH3_32, new Type[] {Type.VARCHAR},
+                Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF);
+        Type type = pickBucketType(bucketNum);
+        CallOperator callHashFuncOp =
+                new CallOperator(FunctionSet.MURMUR_HASH3_32, hashFunc.getReturnType(),
+                        Collections.singletonList(stringCol),
+                        hashFunc);
+
+        // sign of remainder keeps consistency with that of dividend, so divisor should be half of bucketNum.
+        // for an example, when bucketNum=8; the remainder will be
+        // -3, -2, -1, 0, 1, 2, 3, 4, NULL
+        ConstantOperator bucketConstOp = new ConstantOperator(bucketNum / 2, Type.INT);
+        CallOperator modBucketOp =
+                new CallOperator(FunctionSet.MOD, Type.INT, Arrays.asList(callHashFuncOp, bucketConstOp));
+        ScalarOperator op = new CastOperator(type, modBucketOp, true);
+        ScalarOperatorRewriter rewriter = new ScalarOperatorRewriter();
+        return rewriter.rewrite(op, Collections.singletonList(new ReduceCastRule()));
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggOp = input.getOp().cast();
+        OptExpression child = input.inputAt(0);
+
+        Map.Entry<ColumnRefOperator, CallOperator> aggregate = aggOp.getAggregations().entrySet().iterator().next();
+        ColumnRefOperator aggColRef = aggregate.getKey();
+        CallOperator aggCall = aggregate.getValue();
+        boolean countDistinctAgg =
+                aggCall.isDistinct() && aggCall.getFnName().equalsIgnoreCase(FunctionSet.COUNT);
+        if (!countDistinctAgg || aggCall.getChildren().size() > 1 || !aggCall.getChild(0).isColumnRef()) {
+            return Collections.emptyList();
```

Comment:
`checkGroupByCountDistinct()` has do this work?

Reply:
it can be deleted, check method has already done this work.